Add Python bindings for libvillas

al3xa23 opened the pull request:
Python Wrapper for VILLASnode 
Documentation available https://github.com/VILLASframework/documentation/pull/104#issue-2971440658 
Please provide comments what needs to be changed from you point of view :)

Commit:
Build system for python-wrapper

Places the wrapper .so file into the Python lib-dynload folder.
This should always be in the Path when using Python.

The Python Wrapper only builds if necessary build dependencies are
found.

Signed-off-by: Kevin Vu te Laar <vu.te@rwth-aachen.de>

Commit:
Added wrapper dependencies to Dockerfiles

Added python and pybind11 dependencies to Dockerfiles,
except for Fedora minimal, to build the Python-wrapper.

Signed-off-by: Kevin Vu te Laar <vu.te@rwth-aachen.de>

Commit:
Python-Wrapper bindings

Added Python-Wrapper bindings.
Docstrings are still missing.

Signed-off-by: Kevin Vu te Laar <vu.te@rwth-aachen.de>

Commit:
Added Tests for the Python-Wrapper

The tests are not yet integrated into the pipeline and are primarily focused on the signal_v2 node.
Not all functions work as expected.

node_name_short() appears to be print a freed string.

node_stop() and node_restart() do not function properly.
However, they both use the same underlying function to stop a node, which may
be the cause of their failure.

Signed-off-by: Kevin Vu te Laar <vu.te@rwth-aachen.de>

Commit:
node.cpp Bugfix

Fixed logic error.

Threw an error if json_t *json is a json object,
but it should throw an error if that *json weren't a valid json_object.

Signed-off-by: Kevin Vu te Laar <vu.te@rwth-aachen.de>

Commit:
CMake Python-Wrapper test integration, build fix

Integrated the Python-Wrapper test "/test/unit/python_wrapper.py" into
CMake and the CI.
Testing within the Fedora dev container resulted in the issue of missing
permissions.
Manually running:
chmod +x /villas/test/unit/python_wrapper.py
helped. Not sure if this issue requires changes for the CI.

The build integration with CMake had a little issue.
Due to the OPTIONAL flag, pybind11 was never set to found and was therefore
never built even though it may have been found by CMake.

Changed the target for the python-wrapper build from villas_node to
python-wrapper and instead changed its OUTPUT_NAME PROPERTY to
villas_node, so that the Python module can still be imported with:
"import villas_node". This also avoids confusion with CI and CMake
integration.

Signed-off-by: Kevin Vu te Laar <vu.te@rwth-aachen.de>

Commit:
Applied changes as per feedback.

Signed-off-by: Kevin Vu te Laar <vu.te@rwth-aachen.de>

al3xa23 committed:
Merge branch 'master' into python-wrapper

Commit:
Fixed typo

The target is `run-python-wrapper-tests` instead of
`python-wrapper-tests`

Signed-off-by: Kevin Vu te Laar <vu.te@rwth-aachen.de>

Commit:
Explicitly disallowed copy and assignment

cppcheck threw a warning during the pipeline.
Since copy and assignment are not used and should not be used,
this is fine and won't result in any difference.

Calling .copy() in python would have also thrown an error earlier,
as a copyconstructor wasn't explicitly exposed to pybind via bindings.

Signed-off-by: Kevin Vu te Laar <vu.te@rwth-aachen.de>

Commit:
CMakeLists.txt fix

Use find_package(Python3 REQUIRED) to set Python3_EXECUTABLE.

Moved the `REQUIRED` to align with CMake style guidelines.

Signed-off-by: Kevin Vu te Laar <vu.te@rwth-aachen.de>

stv0g commented on tests/unit/CMakeLists.txt:
```diff
@@ -35,6 +35,7 @@ add_custom_target(run-unit-tests
     USES_TERMINAL
 )
 
+find_package(Python3 REQUIRED COMPONENTS Interpreter)
```
Are you sure this is needed? I dont see the variables/targets defined by `find_package` being used in this file.

Normally, such transitive dependencies should be defined using the `PUBLIC` keyword in the `target_link_library`, so they propagate downstream.

stv0g commented on .gitlab-ci.yml:
```diff
@@ -170,6 +170,7 @@ test:unit:
   script:
     - cmake -S . -B build ${CMAKE_OPTS}
     - cmake --build build ${CMAKE_BUILD_OPTS} --target run-unit-tests run-unit-tests-common
+    - cmake --build build ${CMAKE_BUILD_OPTS} --target run-python-wrapper-tests
```
```suggestion
    - cmake --build build ${CMAKE_BUILD_OPTS} --target run-unit-tests run-unit-tests-common run-python-wrapper-tests
```

stv0g commented on clients/CMakeLists.txt:
```diff
@@ -5,4 +5,5 @@
 # SPDX-License-Identifier: Apache-2.0
 
 add_subdirectory(opal-rt/rtlab-asyncip)
+add_subdirectory(python-wrapper)
```
Usually such language support interfaces are called "bindings". Hence also the name "pybind".

I propose to rename this to "python-binding". Its not really wrapping anything in a sense that it builds on top or extends VILLASnode's functionality.

stv0g commented on clients/python-wrapper/villas-python-wrapper.cpp:
```diff
@@ -0,0 +1,246 @@
+/* Python-wrapper.
```
```suggestion
/* Python bindings.
```

stv0g commented on clients/python-wrapper/villas-python-wrapper.cpp:
```diff
@@ -0,0 +1,246 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <jansson.h>
+#include <pybind11/pybind11.h>
+#include <uuid/uuid.h>
+#include <villas/node.hpp>
+#include <villas/sample.hpp>
```
Please separate headers in the following three sections:

- System headers
- Third-party library headers
- VILLASnode headers

```suggestion
#include <jansson.h>
#include <pybind11/pybind11.h>
#include <uuid/uuid.h>

#include <villas/node.hpp>
#include <villas/sample.hpp>
```

stv0g commented on packaging/docker/Dockerfile.debian:
```diff
@@ -51,7 +51,10 @@ RUN apt-get update && \
 		libnice-dev \
-		libmodbus-dev
+		libmodbus-dev \
+		python3 \
```
Depending on `python3` here is not required as its a direct dependency of `python3-dev`.
See: https://debian.pkgs.org/12/debian-main-arm64/python3-dev_3.11.2-1+b1_arm64.deb.html

E.g. we are also not depending on `libmodbus`, and only on `libmodbus-dev`.

stv0g commented on packaging/docker/Dockerfile.debian-multiarch:
```diff
@@ -58,7 +58,10 @@ RUN apt-get update && \
 		libhiredis-dev:${ARCH} \
-		libmodbus-dev:${ARCH}
+		libmodbus-dev:${ARCH} \
+		python3:${ARCH} \
```
Some as above.

stv0g commented on packaging/docker/Dockerfile.ubuntu:
```diff
@@ -59,7 +59,10 @@ RUN apt-get update && \
 		libglib2.0-dev \
-		libcriterion-dev
+		libcriterion-dev \
+		python3 \
```
Same as above.

stv0g commented on clients/python-wrapper/CMakeLists.txt:
```diff
@@ -0,0 +1,39 @@
+# SPDX-FileCopyrightText: 2014-2025 Institute for Automation of Complex Power Systems, RWTH Aachen University
+# SPDX-License-Identifier: Apache-2.0
+
+cmake_minimum_required(VERSION 3.15...3.29)
```
Could we move this to the top-level `CMakeLists.txt`? Or is there a reasoning for having this also here?

Building this as a dedicated project wont work due to the missing `villas` library target which you use below..

KeVteL replied:
addressed in: 6806d2409f5fba50dbed4b01589b20e518bf5787

stv0g commented:
Hi @al3xa23 do you or your student need support in getting this merged?

al3xa23 commented:
Hi, sorry, we need some more time here. If we need help, we come back to you. Thanks :)

Commit:
Renamed wrapper -> binding, LLVM format

As suggested `wrapper` was substituted by `binding`.
The minimum required version and project naming in `CMakeLists.txt`
were removed, as this is not necessary.

Signed-off-by: Kevin Vu te Laar <vu.te@rwth-aachen.de>

Commit:
Changes for pipeline test integration

- Removed unnecessary `find_package()` call in `CMakeLists.txt`.
- Changed naming to match substitution of `wrapper` with `binding`.
- Bumped the version requirement from `3.14` to `3.15` matching the removed minimum in `binding` `CMakeLists.txt`.
- Dropped version range syntax `cmake_minimum_required(A...B)`, introduced
in `3.19`, in favor of a simple minimum requirement of `3.15`.

Signed-off-by: Kevin Vu te Laar <vu.te@rwth-aachen.de>

Commit:
Removed unnecessary dockerfile dependencies

Based on suggestions from PR#884.
Removed the `python` dependency because `python-devel` in Debian and Rocky
also depend on the base `python` package.

Signed-off-by: Kevin Vu te Laar <vu.te@rwth-aachen.de>

al3xa23 committed:
Merge branch 'master' into python-wrapper

Signed-off-by: al3xa23 <140614263+al3xa23@users.noreply.github.com>

Commit:
Binding test formatted with black

Manually formatted the Python binding test with black.
The pipeline formatting resulted in an invalid escape sequence `\m`.

Oddly enough there is no such character sequence within the code.
Especially not at the reported line 108 when looking at:
https://git.rwth-aachen.de/acs/public/villas/node/-/jobs/6248036 .

Signed-off-by: Kevin Vu te Laar <vu.te@rwth-aachen.de>

Commit:
Format Python binding test with black-jupyter

Manually formatted the Python binding test with black-jupyter.
The pipeline formatting resulted in an "invalid escape sequence \m"
error.

Oddly enough there is no such character sequence within the code. Especially not at the reported line 108 when considering:
https://git.rwth-aachen.de/acs/public/villas/node/-/jobs/6248036
https://git.rwth-aachen.de/acs/public/villas/node/-/jobs/6249567

`black-jupyter` seems to enforce stricter formatting rules than plain `black`.
Formatting with plain `black` caused the pipeline to fail, as `black-jupyter` made additional changes.

Signed-off-by: Kevin Vu te Laar <vu.te@rwth-aachen.de>

KeVteL commented on .gitlab-ci.yml:
```diff
@@ -173,7 +173,8 @@ test:unit:
     - cmake -S . -B build ${CMAKE_OPTS}
-    - cmake --build build ${CMAKE_BUILD_OPTS} --target run-unit-tests run-unit-tests-common
+    - export PYTHONPATH=$PYTHONPATH:${PWD}/build/clients/python-binding
+    - cmake --build build ${CMAKE_BUILD_OPTS} --target run-unit-tests run-unit-tests-common run-python-binding-tests
```
regarding:
https://github.com/VILLASframework/node/pull/884#discussion_r2033663485

I am not quite sure whether or not it is better to export the path of the binding `.so` as is done here or
to make use of something like this:
```yaml
paths:
  -build/clients/python-binding
  ```
  
  I am not sure whether or not this would even work.

KeVteL commented:
regarding: https://github.com/VILLASframework/node/pull/884#discussion_r2033673569 https://github.com/VILLASframework/node/pull/884#discussion_r2033673878 https://github.com/VILLASframework/node/pull/884#discussion_r2033674228

I also removed the package `python` from `Dockerfile.fedora`, as `python-devel` also depends on it. b1e9a2f637f3430af9c9f2cb44ef3fdbbfdbac54

stv0g commented on packaging/docker/Dockerfile.rocky:
```diff
@@ -25,7 +25,8 @@ RUN dnf --allowerasing -y install \
 	protobuf-compiler protobuf-c-compiler \
-	clang-tools-extra
+	clang-tools-extra \
+	python-devel
```
Do you mind moving this below to the pybind11-devel package, as for the other Dockerfiles?

stv0g replied:
Hi @KeVteL,

> I realized during a debian install, that the dependency list of the Dockerfiles debian, debian-multiarch and rocky are missing libwebsockets(-dev), which is required to build VILLASnode.

This maybe due to the version of libwebsocket shipped in the distributions package repository to be too old. If a package is not already installed (e.g. by the Dockerfile), we will install the missing dependencies (using the correct newer versions) via the `deps.sh` shell script.

However, I would be surprised if for the particular case of libwebsocket, the distro packages are too old. We should check here.

We also have upgrade some of the base distro versions (e.g. Debian 12 -> 13) in the last months without checking all the other dependencies. It might be that we now have the required versions already provided by the distros.

The exact version requirements are documented here:
- https://villas.fein-aachen.org/docs/node/installation#prerequisites
- https://github.com/VILLASframework/node/blob/9b81d28fe1ce03c40393251837b524087c50b03c/CMakeLists.txt#L95

stv0g commented on tests/unit/CMakeLists.txt:
```diff
@@ -35,5 +35,13 @@ add_custom_target(run-unit-tests
     USES_TERMINAL
 )
 
+add_custom_target(run-python-binding-tests
```
I am not so sure if this is really a unit-test, as its not testing individual and isolated parts of the code (e.g. functions).

Instead the Python test seems a bit more elaborate.

What do you think about adding it as an integration test? Or maybe as a dedicated test in itself?

I am interested in your point of view here :)

stv0g reviewed (changes requested):
Great work :) I have only minor points regarding packaging..

Please let me know if you need some support on this.

Commit:
Align python-devel package for consistency

Moved `python-devel` to align with the Ubuntu and Debian Dockerfile.
The Fedora one remains the exception.

Signed-off-by: Kevin Vu te Laar <vu.te@rwth-aachen.de>

al3xa23 commented:
Hi @stv0g, can you please give some (final?) comment here so that we can proceed? Thanks :)

Commit:
Moved clients/python-binding/ to python/binding/

Signed-off-by: Kevin Vu te Laar <vu.te@rwth-aachen.de>

Commit:
python-binding unit test split, socket fixes

Unit test portion split.
Previously commented-out tests are now active.
Fixes to the socket node resolved previous issues during testing.
These arised when C-API functions were called through the bindings in likely unintended ways,
causing undefined behavior (unexpected return values), segfaults or other memory related crashes.

Fixes:
- Socket descriptor (`sd`) is now initialized with -1
- `in` and `out` buffers are initialized with `nullptr`
- Defensive deallocation and cleanup of buffers:
  * Check buffers before freeing
  * Zero memory before freeing
  * Set pointers to `nullptr` after freeing

Signed-off-by: Kevin Vu te Laar <vu.te@rwth-aachen.de>

Commit:
unit tests: add test and enable test discovery

Unit Tests in `tests/unit/python/` are now executed automatically as long as
their filenames start with the prefix `test`.

The binding unit test was renamed/moved from
`tests/unit/python_binding.py` to
`tests/unit/python/test_python_binding.py`.

Node implementations return `-1` for functions that are not implemented.
This is now covered for the `node_reverse()` function of the `socket`
node.

Signed-off-by: Kevin Vu te Laar <vu.te@rwth-aachen.de>

Commit:
integration tests: added and enable test discovery

integration tests are now:
- split from unit tests
- improved to cover more cases
- automatically executed upon discovery

Integration tests must have the filename prefix `test` to be discovered.

Signed-off-by: Kevin Vu te Laar <vu.te@rwth-aachen.de>

Commit:
python binding: add wrapper and binding tweaks

- Added a python-wrapper for the python-binding:
  * Automatic memory management
  * Restricted access and standardized sample handling
  * Added documentation via docstrings

Binding Tweaks:
  * Support bulk allocation
  * Support writing sample slices/blocks

Signed-off-by: Kevin Vu te Laar <vu.te@rwth-aachen.de>

Commit:
binding tests: add binding wrapper tests

- CI/CD integration of binding wrapper tests.
- Add unit and integration tests for binding wrapper.

Signed-off-by: Kevin Vu te Laar <vu.te@rwth-aachen.de>

Commit:
binding tests: CI fix and reformat

- Updated PYTHONPATH to match directory structure.
- Python integration tests are now executed.
- Binding related `.py`-files now adhere to 79-character line length.

Signed-off-by: Kevin Vu te Laar <vu.te@rwth-aachen.de>

Commit:
binding wrapper: CI/build fix, add binding stubs

CI:
- Python unit and integration tests have their own CI test now.
- Unit and integration test pipelines reverted to pre-binding related
  changes. Binding need workarounds to comply with packaging, CI
  and building compatibility.
- Format fixes.
- `binding.py` (binding wrapper) and bindings have stubs now.

CMake build:
- Naming is more straight forward.
- Bindings install into `pythonX.Y/site-packages/villas/node/`.
  Should be compatible with a potential pypi-package related to the
  wrapper binding.

Signed-off-by: Kevin Vu te Laar <vu.te@rwth-aachen.de>

Commit:
binding stubs: add missing SPDX header

Also reformatted the stubs with pre-commit run on the
`python:3.12.10-slim-bookworm` image that is also used in the pipeline.

Running pre-commit within the Fedora dev image has different behavior.

Signed-off-by: Kevin Vu te Laar <vu.te@rwth-aachen.de>

Commit:
python binding: install and fix CI

Install:

- As of Python 3.3+, `__init__.py` is not required in site-package
  folders. Those folders are seen as module by default.
- Helps with file ownership when installing the binding
  wrapper via pip as part of the villas-node package and the bindings from the codebase.

CI:

- Fixed wrong variable name in the CI script.
- .so file should be found and symlinked properly into `/python/villas/node` for testing purposes.

Signed-off-by: Kevin Vu te Laar <vu.te@rwth-aachen.de>

Commit:
Merge branch 'master' into python-wrapper

Signed-off-by: Kevin Vu te Laar <vu.te@rwth-aachen.de>

Commit:
Rework Binding Wrapper

Binding rework to a class design.

Node:
- Stores SamplesArray, Config, cpp-Node handle.
- Automatically deals with SamplesArray sizing and (de-)allocation of
  samples.
- Removed `node_netem_fds()` and `node_poll_fds`.
- `sample_pack()` and `sample_unpack` can only read/write samples to a
  SamplesArray within a node.
- Added `sample_details()` for inspecting sample content.
- Added `_SamplesArray` helper class to enable more pythonic syntax.

node_capi.cpp, node.h:
- Aligned `sequence` parameter type in `sample_pack()` and `sample_unpack()`
  with `sequence` field of `struct Sample` (uint64_t)

Signed-off-by: Kevin Vu te Laar <vu.te@rwth-aachen.de>

Commit:
Update stubs to match changes

Signed-off-by: Kevin Vu te Laar <vu.te@rwth-aachen.de>

Commit:
Update tests to match binding changes

Updated function names and evoking syntax.
Added test for:
  - Automatic SamplesArray sizing
  - `sample_pack()`, `sample_unpack()` accessible via `pack_from()`,
   `unpack_to`
  - `samples_detail()`

Signed-off-by: Kevin Vu te Laar <vu.te@rwth-aachen.de>

Commit:
Merge branch 'master' into python-wrapper

Signed-off-by: Kevin Vu te Laar <vu.te@rwth-aachen.de>

Commit:
Enforce pre-commit formatting

Reformat python files to adhere to the updated 90-character
line-length setting in pyproject.toml.

Signed-off-by: Kevin Vu te Laar <vu.te@rwth-aachen.de>

Commit:
Replace C with C++ casting, sample pack/unpack fix

Replace C-style casts with C++-style casts.

`sample_unpack()` always unpacks source sample timespec instead of default or
user-specified values.

`sample_pack()` uses user-specified if provided; otherwise falls back to source
sample timespec when packing from a sample, or to the system clock when
packing from a list.

Modified stubs to reflect changes.

Signed-off-by: Kevin Vu te Laar <vu.te@rwth-aachen.de>

Commit:
fix(ci): prevent mypy package root ambiguity

Mypy was detecting the same files under multiple module names
due to ambiguous package roots. Explicit package bases prevent ambiguity.

Signed-off-by: Kevin Vu te Laar <vu.te@rwth-aachen.de>

leonardocarreras commented on clients/python-binding/CMakeLists.txt:
Hi, I know it is a year later... But some extra comments here... In the end my understanding is that you need to use the PyPI tools to make your life easier
- For the case of how to do this with the tools, I would like to point to [cibuildwheel](https://cibuildwheel.pypa.io/en/stable/)
    - If you need an example of this, you can check DPsim [here](https://github.com/sogno-platform/dpsim/blob/b63e7153f2cc94ad8844e1110efeb7e2ac9c13e9/.github/workflows/publish_to_pypi.yaml#L41)
    - You need to provide your own Containerfile like [here](https://github.com/sogno-platform/dpsim/blob/master/packaging/Docker/Dockerfile.manylinux)
- About the CMake compilation, this mirrors a bit the way villas does, so you need the villas libraries and the python code to get it packaged, plus the tooling will take care to include the libraries inside the package
    - About the code of the repo and how to compile, that is not a problem once you use the tooling and you change the way that is done, you can use a [pyproject.toml](https://github.com/sogno-platform/dpsim/blob/master/pyproject.toml) that runs a [setup.py](https://github.com/sogno-platform/dpsim/blob/master/setup.py) which does the CMake compilation, and select the respective targets that will be part of the package
    - You need to make sure that libraries are available in the container, and that you include the python code like [here](https://github.com/sogno-platform/dpsim/blob/b63e7153f2cc94ad8844e1110efeb7e2ac9c13e9/pyproject.toml#L58)

leonardocarreras commented on python/binding/capi_python_binding.cpp:
```diff
@@ -0,0 +1,404 @@
+/* Python-binding.
+ *
+ * Author: Kevin Vu te Laar <vu.te@rwth-aachen.de>
+ * SPDX-FileCopyrightText: 2014-2025 Institute for Automation of Complex Power
```
2025? I think this file is new

leonardocarreras replied:
do not break this line

leonardocarreras commented on python/binding/capi_python_binding.cpp:
```diff
@@ -0,0 +1,404 @@
+ * @param villas_node   Name of the module to be bound
+ * @param m             Access variable for modifying the module code
+ */
+PYBIND11_MODULE(python_binding, m) {
```
any specific reason why all is prefixed with `node_`

leonardocarreras commented on python/binding/capi_python_binding.cpp:
```diff
@@ -0,0 +1,404 @@
+
+struct timespec ns_to_timespec(int64_t time_ns) {
+  struct timespec ts;
+  ts.tv_nsec = time_ns / 1'000'000'000LL;
```
this looks flipped, isnt it nsec the reminder, and sec the division?

ts.tv_nsec = time_ns / 1'000'000'000LL;
ts.tv_sec  = time_ns % 1'000'000'000LL;

leonardocarreras commented on python/binding/capi_python_binding.cpp:
```diff
@@ -0,0 +1,404 @@
+    if (this->smps) {
+      unsigned int i = 0;
+      for (; i < len; ++i) {
+        sample_decref(smps[i]);
```
no check for` smps[i]` is null

leonardocarreras commented on python/binding/capi_python_binding.cpp:
```diff
@@ -0,0 +1,404 @@
+      .def(py::init<unsigned int>(), py::arg("len"))
+      .def("__getitem__",
+           [](SamplesArray &a, unsigned int idx) {
+             assert(idx < a.size() && "Index out of bounds");
```
better throw an exception

leonardocarreras commented on python/binding/capi_python_binding.cpp:
```diff
@@ -0,0 +1,404 @@
+           })
+      .def("__setitem__",
+           [](SamplesArray &a, unsigned int idx, void *smp) {
+             assert(idx < a.size() && "Index out of bounds");
```
better throw an exception

leonardocarreras commented on python/binding/capi_python_binding.cpp:
```diff
@@ -0,0 +1,404 @@
+            ts_received_ns ? ns_to_timespec(*ts_received_ns) : time_now();
+
+        unsigned values_len = values.size();
+        double cvalues[values_len];
```
can we use std::vector here?

leonardocarreras commented on lib/node_capi.cpp:
```diff
@@ -169,7 +169,7 @@ vsample *sample_pack(unsigned seq, struct timespec *ts_origin,
 }
 
-void sample_unpack(vsample *s, unsigned *seq, struct timespec *ts_origin,
+void sample_unpack(vsample *s, uint64_t *seq, struct timespec *ts_origin,
```
this breaks the C API, is this needed?

leonardocarreras commented on lib/node_capi.cpp:
```diff
@@ -152,12 +152,12 @@ unsigned sample_length(vsample *s) {
 }
 
-vsample *sample_pack(unsigned seq, struct timespec *ts_origin,
+vsample *sample_pack(uint64_t *seq, struct timespec *ts_origin,
```
this breaks the C API, is this needed?

leonardocarreras commented on include/villas/node.h:
```diff
@@ -67,11 +68,11 @@ unsigned sample_length(vsample *smp);
 void sample_decref(vsample *smp);
 
-vsample *sample_pack(unsigned seq, struct timespec *ts_origin,
+vsample *sample_pack(uint64_t *seq, struct timespec *ts_origin,
```
this breaks the C API, is this needed?

leonardocarreras commented on include/villas/node.h:
```diff
@@ -67,11 +68,11 @@ unsigned sample_length(vsample *smp);
                      double *values);
 
-void sample_unpack(vsample *s, unsigned *seq, struct timespec *ts_origin,
+void sample_unpack(vsample *s, uint64_t *seq, struct timespec *ts_origin,
```
this breaks the C API, is this needed?

leonardocarreras commented on include/villas/node.h:
```diff
@@ -6,6 +6,7 @@
  */
 
 #pragma once
+#include <cstdint>
```
can we use `stdint.h `?

leonardocarreras commented on python/binding/capi_python_binding.cpp:
```diff
@@ -0,0 +1,404 @@
+ *
+ * Author: Kevin Vu te Laar <vu.te@rwth-aachen.de>
+ * SPDX-FileCopyrightText: 2014-2025 Institute for Automation of Complex Power
+ * Systems, RWTH Aachen University SPDX-License-Identifier: Apache-2.0
```
the SPDX-License should be in a new line

leonardocarreras commented on tests/integration/CMakeLists.txt:
```diff
@@ -18,4 +18,12 @@ add_custom_target(run-integration-tests
+    USES_TERMINAL
+)
+
+add_dependencies(run-tests run-integration-tests run-python-integration-tests)
```
missing python_bindings dependency? will fail if that is not there

leonardocarreras commented on .gitlab-ci.yml:
```diff
@@ -182,6 +182,28 @@ test:cppcheck:
       - cppcheck.log
     expose_as: cppcheck
 
+test:python_unit_integration:
```
this need to be converted to cibuildwheels and github action

leonardocarreras commented on tests/unit/CMakeLists.txt:
```diff
@@ -35,5 +35,13 @@ add_custom_target(run-unit-tests
+add_custom_target(run-python-unit-tests
+    COMMAND
+        python3 -m unittest discover ${CMAKE_CURRENT_SOURCE_DIR}
+    DEPENDS
```
`DEPENDS` does not work as `add_dependencies`, is that what you meant?

leonardocarreras commented on tests/unit/python/test_python_binding.py:
```diff
@@ -0,0 +1,239 @@
+"""
+Author: Kevin Vu te Laar <vu.te@rwth-aachen.de>
+SPDX-FileCopyrightText: 2014-2025 Institute for Automation of Complex Power Systems, RWTH Aachen University
```
2025?

leonardocarreras commented on tests/unit/python/test_binding_wrapper.py:
```diff
@@ -0,0 +1,241 @@
+"""
+Author: Kevin Vu te Laar <vu.te@rwth-aachen.de>
+SPDX-FileCopyrightText: 2014-2025 Institute for Automation of Complex Power Systems, RWTH Aachen University
```
2025?

leonardocarreras commented on tests/unit/python/test_binding_wrapper.py:
```diff
@@ -0,0 +1,241 @@
+Author: Kevin Vu te Laar <vu.te@rwth-aachen.de>
+SPDX-FileCopyrightText: 2014-2025 Institute for Automation of Complex Power Systems, RWTH Aachen University
+SPDX-License-Identifier: Apache-2.0
+"""  # noqa: E501
```
what is noqa: E501

leonardocarreras commented on tests/integration/CMakeLists.txt:
```diff
@@ -18,4 +18,12 @@ add_custom_target(run-integration-tests
+add_custom_target(run-python-integration-tests
+    COMMAND
+        python3 -m unittest discover ${CMAKE_CURRENT_SOURCE_DIR}
+    DEPENDS
```
DEPENDS vs add_dependencies

leonardocarreras commented on tests/integration/python/test_python_binding.py:
```diff
@@ -0,0 +1,287 @@
+"""
+Author: Kevin Vu te Laar <vu.te@rwth-aachen.de>
+SPDX-FileCopyrightText: 2014-2025 Institute for Automation of Complex Power Systems, RWTH Aachen University
```
2025?

leonardocarreras commented on tests/integration/python/test_binding_wrapper.py:
```diff
@@ -0,0 +1,305 @@
+            self.fail(f"err: {e}")
+
+
+test_node_config = {
```
can we store the configs somewhere? they also seem to be quite similar one to each other

leonardocarreras commented on CMakeLists.txt:
Add a WITH_PYTHON_BINDINGS option

leonardocarreras commented on .gitlab-ci.yml:
```diff
@@ -164,7 +164,7 @@ test:python:
     - black --line-length=90 --extend-exclude=".*(\\.pyi|_pb2.py)$" --check .
     - flake8 --max-line-length=90 --extend-exclude="*.pyi,*_pb2.py" .
-    - mypy .
+    - mypy --explicit-package-bases .
```
pybind provides a way to generate stubs, would we want to do this still with mypy?

leonardocarreras reviewed (commented):
Hi @KeVteL! Thanks for your PR, we just took a new impulse and want to help to get our contributions merged... So sorry already for the delays...

I have some comments, many are questions, many are preferences... Is not meant to be something definitive or a disagreement, but a starting point to think about the direction...
- I have the feeling that we can use CMakeExtension or scikit-build helps with getting built all together
- If the previous is true, you can use cibuildtools so then this is easier
- I prefer pybind11-stubgen to hand-written pyi, I think even now some of the stubs are out of sync with their functions/classes
- Using auditwheel can help patch the package and include the libraries, so that is also not a problem theoretically
- What exactly to bind is still to me a question but I would say is more on the direction of villas/node.hpp, I could be missing something... Many new functions and classes somehow reimplement the logic of the c/c++ code but in python... Maybe this is something to discuss more... But I have the feeling that we could avoid this layering that duplicates, ref counters and pack/unpack...
- Node is somehow declared in many places as of the state now, we could end up shadowing one with each other unadvertedly in my opinion
- Some more nitpicky things like the years and lines on SPDX (new files could have only 2025 I guess, unless something different is asked in the contributing guidelines)
- Maybe configs can be separated from their tests when they are too similar, to avoid duplication


I think that it would be nice to know your opinions and takes about those items. Thanks in advance!
